internal/pkg/certificates/certificates.go

```diff
@@ -0,0 +1,322 @@
+		}
+		return nil
+	})
+	if err != nil {
+		err = errors.Join(fmt.Errorf("failed to update the validating webhooks"), err)
+	}
+	return err
+}
+
+func updateMutatingWebhooks(ctx context.Context, k8s client.Client, namespace string, caBundle []byte) error { //nolint:dupl
+	err := retry.OnError(retry.DefaultBackoff, func(err error) bool {
+		return apierrors.IsConflict(err) || apierrors.IsNotFound(err)
+	}, func() error {
+		mutatingWebhookConfig := &admissionregistrationv1.MutatingWebhookConfiguration{}
+		if err := k8s.Get(ctx, client.ObjectKey{Name: constants.ControllerMutatingWebhookName, Namespace: namespace}, mutatingWebhookConfig); err != nil {
+			return errors.Join(errors.New("cannot get controller mutating webhooks"), err)
+		}
+		patchMutatingWebhookConfig := mutatingWebhookConfig.DeepCopy()
+		for i := range patchMutatingWebhookConfig.Webhooks {
+			patchMutatingWebhookConfig.Webhooks[i].ClientConfig.CABundle = caBundle
+		}
+		if err := k8s.Patch(ctx, patchMutatingWebhookConfig, client.MergeFrom(mutatingWebhookConfig)); err != nil {
+			return errors.Join(fmt.Errorf("cannot patch the mutating webhook: %q ", mutatingWebhookConfig.Name), err)
+		}
+		return nil
+	})
+	if err != nil {
+		err = errors.Join(fmt.Errorf("failed to update the mutating webhooks"), err)
+	}
+	return err
+}
+
+// updateWebhooksCaBundle updates the webhook configuration to use the given
+// caBundle to validate the connections. It updates both validating and
+// mutating webhooks
+func updateWebhooksCaBundle(ctx context.Context, k8s client.Client, namespace string, caBundle []byte) error {
+	logger.Info("update webhook caBundle")
+	// Controller webhooks names used here should match what is used in the helm chart installation
+	if err := updateValidationWebhooks(ctx, k8s, namespace, caBundle); err != nil {
+		return errors.Join(fmt.Errorf("cannot update validation webhooks"), err)
+	}
+
+	if err := updateMutatingWebhooks(ctx, k8s, namespace, caBundle); err != nil {
+		return errors.Join(fmt.Errorf("cannot update mutating webhooks"), err)
+	}
+
+	return nil
+}
+
+// SetupCA make sure that the Root CA and the certificate used by the webhook
+// server is created.
+//
+// In case the Root CA is missing, this function will follow the below steps:
+//  1. If the root CA is missing, create it and store certificate and key in a
+//     secret
+//  2. If the Root CA is created, update the webhooks to use root CA to
+//     validate connections
+//  3. If root CA is created or webhook server certificate is missing, create a
+//     certificate to be used by the webhook server and store
+//
+// it in a secret
+//  4. Wait until the webhooks server certificates are mount in the local
+//     filesystem.
+//
+// It returns an error if something wrong happen in any of the step mentioned
+// above
+func SetupCA(ctx context.Context, k8sClient client.Client, deploymentsNamespace string, controllerWebhookServiceName string) error {
```

Comment:
nit(readability): we could move the public function on top of the file so it's easier to understand the public interfaces.